tests/simulation/test_cluster.hpp

```diff
@@ -0,0 +1,251 @@
+}
+
+void RunMachine(MachineManager<SimulatorTransport> mm) { mm.Run(); }
+
+void WaitForShardsToInitialize(CoordinatorClient<SimulatorTransport> &coordinator_client) {
+  // Call coordinator client's read method for GetShardMap and keep
+  // reading it until the shard map contains proper replicas for
+  // each shard in the label space.
+
+  while (true) {
+    GetShardMapRequest req{};
+    CoordinatorReadRequests read_req = req;
+    auto read_res = coordinator_client.SendReadRequest(read_req);
+    if (read_res.HasError()) {
+      // timed out
+      continue;
+    }
+    auto response_result = read_res.GetValue();
+    auto response = std::get<GetShardMapResponse>(response_result);
+    auto shard_map = response.shard_map;
+
+    if (shard_map.ClusterInitialized()) {
+      spdlog::info("cluster stabilized - beginning workload");
+      return;
+    }
+  }
+}
+
+ShardMap TestShardMap(int n_splits, int replication_factor) {
+  ShardMap sm{};
+
+  const std::string label_name = std::string("test_label");
+
+  // register new properties
+  const std::vector<std::string> property_names = {"property_1", "property_2"};
+  const auto properties = sm.AllocatePropertyIds(property_names);
+  const auto property_id_1 = properties.at("property_1");
+  const auto property_id_2 = properties.at("property_2");
+  const auto type_1 = memgraph::common::SchemaType::INT;
+  const auto type_2 = memgraph::common::SchemaType::INT;
+
+  // register new label space
+  std::vector<SchemaProperty> schema = {
+      SchemaProperty{.property_id = property_id_1, .type = type_1},
+      SchemaProperty{.property_id = property_id_2, .type = type_2},
+  };
+
+  std::optional<LabelId> label_id = sm.InitializeNewLabel(label_name, schema, replication_factor, sm.shard_map_version);
+  RC_ASSERT(label_id.has_value());
+
+  // split the shard at N split points
+  for (int64_t i = 1; i < n_splits; ++i) {
+    const auto key1 = memgraph::storage::v3::PropertyValue(i);
+    const auto key2 = memgraph::storage::v3::PropertyValue(0);
+
+    const auto split_point = {key1, key2};
+
+    const bool split_success = sm.SplitShard(sm.shard_map_version, label_id.value(), split_point);
+
+    RC_ASSERT(split_success);
+  }
+
+  return sm;
+}
+
+void ExecuteOp(msgs::ShardRequestManager<SimulatorTransport> &shard_request_manager,
+               std::set<CompoundKey> &correctness_model, CreateVertex create_vertex) {
+  const auto key1 = memgraph::storage::v3::PropertyValue(create_vertex.key);
+  const auto key2 = memgraph::storage::v3::PropertyValue(create_vertex.value);
+
+  std::vector<msgs::Value> primary_key = {msgs::Value(int64_t(create_vertex.key)),
+                                          msgs::Value(int64_t(create_vertex.value))};
+
+  if (correctness_model.contains(std::make_pair(create_vertex.key, create_vertex.value))) {
+    // TODO(tyler) remove this early-return when we have properly handled setting non-unique vertexes
+    return;
+  }
```

Comment:
Here it is quite confusing for me why both `create_vertex.key` and `create_vertex.value` is used to create a key. Maybe try to rename the members of `CreateVertex` to explain their purpose better.

Reply:
I've renamed the members to be a bit more clear